bundles/org.openhab.binding.millheat/src/main/java/org/openhab/binding/millheat/internal/handler/MillheatAccountHandler.java

```diff
@@ -0,0 +1,449 @@
+                    success = true;
+                } catch (MillheatCommunicationException e) {
+                    if (AbstractResponse.ERROR_CODE_ACCESS_TOKEN_EXPIRED == e.getErrorCode()
+                            || AbstractResponse.ERROR_CODE_INVALID_SIGNATURE == e.getErrorCode()) {
+                        doLogin();
+                    }
+                }
+            }
+            if (!success) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            }
+        }
+    }
+
+    private void updateThingStatuses() {
+        List<Thing> subThings = getThing().getThings();
+        for (Thing thing : subThings) {
+            ThingHandler handler = thing.getHandler();
+            if (handler != null) {
+                MillheatBaseThingHandler mHandler = (MillheatBaseThingHandler) handler;
+                mHandler.updateState(model);
+            }
+        }
+    }
+
+    private Request buildLoggedInRequest(AbstractRequest req)
+            throws NoSuchAlgorithmException, UnsupportedEncodingException {
+        String nonce = getRandomString(NUM_NONCE_CHARS);
+        String timestamp = String.valueOf(System.currentTimeMillis() / 1000);
+        String signatureBasis = REQUEST_TIMEOUT + timestamp + nonce + token;
+        String signature = DigestUtils.shaHex(signatureBasis);
+        String reqJson = gson.toJson(req);
+
+        Request request = httpClient.newRequest(API_ENDPOINT_2 + req.getRequestUrl());
+
+        return addStandardHeadersAndPayload(request, req).header("X-Zc-Timestamp", timestamp)
+                .header("X-Zc-Timeout", REQUEST_TIMEOUT).header("X-Zc-Nonce", nonce).header("X-Zc-User-Id", userId)
+                .header("X-Zc-User-Signature", signature).header("X-Zc-Content-Length", "" + reqJson.length());
+
+    }
+
+    private Request addStandardHeadersAndPayload(Request req, AbstractRequest payload)
+            throws UnsupportedEncodingException {
+        requestLogger.listenTo(req);
+
+        return req.header("Connection", "Keep-Alive").header("X-Zc-Major-Domain", "seanywell")
+                .header("X-Zc-Msg-Name", "millService").header("X-Zc-Sub-Domain", "milltype").header("X-Zc-Seq-Id", "1")
+                .header("X-Zc-Version", "1").method(HttpMethod.POST).timeout(5, TimeUnit.SECONDS)
+                .content(new BytesContentProvider((gson.toJson(payload)).getBytes("UTF-8")), CONTENT_TYPE);
+    }
+
+    public void updateRoomTemperature(String roomId, Command command, ModeType mode) {
+        Optional<Home> optionalHome = model.findHomeByRoomId(roomId);
+        Optional<Room> optionalRoom = model.findRoomById(roomId);
+        if (optionalHome.isPresent() && optionalRoom.isPresent()) {
+            SetRoomTempRequest req = new SetRoomTempRequest(optionalHome.get(), optionalRoom.get());
+            if (command instanceof QuantityType<?>) {
+                int newTemp = (int) ((QuantityType<?>) command).longValue();
```

Comment:
The unit in the QuantityType can be Farhenheit. So first make sure the value is in the unit you need than get the value:
```suggestion
                int newTemp = command.as(QuantityType.class).toUnit(SIUnits.CELCIUS).intValue();
```

This same comment applies to any place where you get the value from a QuantityType object.